File: ts/y402/src/core/utils/server/html.ts
import { z } from "zod";
import { PAYWALL_TEMPLATE } from "./template";
import { X402_VERSION } from "../../types/constants";
import { PaymentRequirementsSchema, type PaymentRequirements } from "../../types/requirements";
import { type PaywallConfig } from "../../types/paywall";


// Shape of the window.x402 configuration
export const X402ConfigSchema = z.object({
    amount: z.number(),
    paymentRequirements: PaymentRequirementsSchema.array(),
    testnet: z.boolean(),
    currentUrl: z.string(),
    error: z.string(),
    x402_version: z.string(),
    cdpClientKey: z.string(),
    appName: z.string(),
    appLogo: z.string(),
    sessionTokenEndpoint: z.string(),
});


export type X402Config = z.infer<typeof X402ConfigSchema>;


// Allow TypeScript to know about window.x402
declare global {
    interface Window {
        x402?: X402Config;
    }
}


/**
 * Create x402 configuration object from payment requirements.
 * @param error The error to show.
 * @param paymentRequirements The list of supported requirements.
 * @param paywallConfig Optional configuration for the CDP paywall.
 * @returns The final config to use.
 */
export function createX402Config(
    error: string,
    paymentRequirements: PaymentRequirements[],
    paywallConfig?: PaywallConfig | null,
): X402Config {
    const requirements = paymentRequirements[0] ?? null;

    let displayAmount = 0;
    let currentUrl = "";
    let testnet = true;

    if (requirements) {
        // Convert atomic amount back to USD (assuming USDC with 6 decimals)
        const rawAmount = Number(
            // max_amount_required in Python; adapt to your field name if different
            (requirements as any).max_amount_required ?? requirements.maxAmountRequired,
        );

        if (Number.isFinite(rawAmount)) {
            displayAmount = rawAmount / 1_000_000; // USDC has 6 decimals
        } else {
            displayAmount = 0;
        }

        currentUrl = requirements.resource ?? "";
        testnet = requirements.network === "base-sepolia";
    }

    // Get paywall config values or defaults
    const config = paywallConfig ?? ({} as PaywallConfig);

    const baseConfig: X402Config = X402ConfigSchema.parse({
        amount: displayAmount,
        paymentRequirements,
        testnet,
        currentUrl,
        error,
        x402_version: X402_VERSION,
        cdpClientKey: (config as any).cdp_client_key ?? "",
        appName: (config as any).app_name ?? "",
        appLogo: (config as any).app_logo ?? "",
        sessionTokenEndpoint: (config as any).session_token_endpoint ?? "",
    });

    // Validate/normalize with Zod at runtime (optional but keeps things strict)
    return X402ConfigSchema.parse(baseConfig);
}

/**
 * Injects payment requirements into HTML as JavaScript variables.
 * @param htmlContent The HTML content to insert.
 * @param error Error message to display.
 * @param paymentRequirements List of payment requirements.
 * @param paywallConfig Optional paywall UI configuration.
 * @returns The final contents.
 */
export function injectPaymentData(
    htmlContent: string,
    error: string,
    paymentRequirements: PaymentRequirements[],
    paywallConfig?: PaywallConfig | null,
): string {
    // Create x402 configuration object
    const x402Config = createX402Config(error, paymentRequirements, paywallConfig);

    // Create the configuration script (matching the Python/TS pattern)
    const logOnTestnet = x402Config.testnet
        ? "console.log('Payment requirements initialized:', window.x402);"
        : "";

    const configScript = `
  <script>
    window.x402 = ${JSON.stringify(x402Config)};
    ${logOnTestnet}
  </script>`;

    // Inject the configuration script into the head
    return htmlContent.replace("</head>", `${configScript}\n</head>`);
}


/**
 * Load paywall HTML and inject payment data.
 * @param error Error message to display.
 * @param paymentRequirements List of payment requirements.
 * @param paywallConfig Optional paywall UI configuration.
 * @returns Complete HTML with injected payment data.
 */
export function getPaywallHtml(
    error: string,
    paymentRequirements: PaymentRequirements[],
    paywallConfig?: PaywallConfig | null,
): string {
    return injectPaymentData(PAYWALL_TEMPLATE, error, paymentRequirements, paywallConfig);
}

File: python/src/y402/core/utils/html.py
import json
from typing import List, Optional, Dict, Any
from .template import PAYWALL_TEMPLATE
from ..types.constants import X402_VERSION
from ..types.paywall import PaywallConfig
from ..types.requirements import PaymentRequirements


def create_x402_config(
    error: str,
    payment_requirements: List[PaymentRequirements],
    paywall_config: Optional[PaywallConfig] = None,
) -> Dict[str, Any]:
    """
    Create x402 configuration object from payment requirements.
    """

    requirements = payment_requirements[0] if payment_requirements else None
    display_amount = 0
    current_url = ""
    testnet = True

    if requirements:
        # Convert atomic amount back to USD (assuming USDC with 6 decimals)
        try:
            display_amount = (
                    float(requirements.max_amount_required) / 1000000
            )  # USDC has 6 decimals
        except (ValueError, TypeError):
            display_amount = 0

        current_url = requirements.resource or ""
        testnet = requirements.network == "base-sepolia"

    # Get paywall config values or defaults
    config = paywall_config or {}

    # Create the window.x402 configuration object
    return {
        "amount": display_amount,
        "paymentRequirements": [
            req.model_dump(by_alias=True) for req in payment_requirements
        ],
        "testnet": testnet,
        "currentUrl": current_url,
        "error": error,
        "x402Version": X402_VERSION,
        "cdpClientKey": config.get("cdp_client_key", ""),
        "appName": config.get("app_name", ""),
        "appLogo": config.get("app_logo", ""),
        "sessionTokenEndpoint": config.get("session_token_endpoint", ""),
    }


def inject_payment_data(
    html_content: str,
    error: str,
    payment_requirements: List[PaymentRequirements],
    paywall_config: Optional[PaywallConfig] = None,
) -> str:
    """
    Injects payment requirements into HTML as JavaScript variables.

    Args:
        html_content: The HTML content to insert.
        error: Error message to display.
        payment_requirements: List of payment requirements.
        paywall_config: Optional paywall UI configuration.

    Returns:
        The final contents.
    """

    # Create x402 configuration object
    x402_config = create_x402_config(error, payment_requirements, paywall_config)

    # Create the configuration script (matching TypeScript pattern)
    log_on_testnet = (
        "console.log('Payment requirements initialized:', window.x402);"
        if x402_config["testnet"]
        else ""
    )

    config_script = f"""
  <script>
    window.x402 = {json.dumps(x402_config)};
    {log_on_testnet}
  </script>"""

    # Inject the configuration script into the head (same as TypeScript)
    return html_content.replace("</head>", f"{config_script}\n</head>")


def get_paywall_html(
    error: str,
    payment_requirements: List[PaymentRequirements],
    paywall_config: Optional[PaywallConfig] = None,
) -> str:
    """
    Load paywall HTML and inject payment data.

    Args:
        error: Error message to display.
        payment_requirements: List of payment requirements.
        paywall_config: Optional paywall UI configuration.

    Returns:
        Complete HTML with injected payment data.
    """

    return inject_payment_data(
        PAYWALL_TEMPLATE, error, payment_requirements, paywall_config
    )
